File: game/src/main/java/com/exjets/aetheria/world/World.java
package com.exjets.aetheria.world;

import com.exjets.aetheria.npc.Npc;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/** The tile grid plus everything standing on it. */
public class World {

    private final String name;
    private final int width;
    private final int height;
    private final TileType[][] tiles;
    private final Map<Position, WorldObject> objects = new HashMap<>();
    private final List<Npc> npcs = new ArrayList<>();
    private final List<GroundItem> groundItems = new ArrayList<>();
    private final Position spawn;

    public World(String name, int width, int height, TileType[][] tiles, Position spawn) {
        this.name = name;
        this.width = width;
        this.height = height;
        this.tiles = tiles;
        this.spawn = spawn;
    }

    public String name() {
        return name;
    }

    public int width() {
        return width;
    }

    public int height() {
        return height;
    }

    public Position spawn() {
        return spawn;
    }

    public boolean inBounds(int x, int y) {
        return x >= 0 && y >= 0 && x < width && y < height;
    }

    public TileType tile(int x, int y) {
        return inBounds(x, y) ? tiles[y][x] : TileType.WALL;
    }

    public TileType tile(Position position) {
        return tile(position.x(), position.y());
    }

    public boolean isWalkable(int x, int y) {
        if (!inBounds(x, y) || !tiles[y][x].walkable()) {
            return false;
        }
        WorldObject object = objects.get(new Position(x, y));
        return object == null || !object.blocksMovement();
    }

    public boolean isWalkable(Position position) {
        return isWalkable(position.x(), position.y());
    }

    public void addObject(WorldObject object) {
        objects.put(object.position(), object);
    }

    public void removeObject(Position position) {
        objects.remove(position);
    }

    public WorldObject objectAt(Position position) {
        return objects.get(position);
    }

    public Map<Position, WorldObject> objects() {
        return objects;
    }

    public List<Npc> npcs() {
        return npcs;
    }

    public void addNpc(Npc npc) {
        npcs.add(npc);
    }

    public Npc npcAt(Position position) {
        for (Npc npc : npcs) {
            if (npc.isVisible() && npc.position().equals(position)) {
                return npc;
            }
        }
        return null;
    }

    public List<GroundItem> groundItems() {
        return groundItems;
    }

    public void addGroundItem(GroundItem item) {
        groundItems.add(item);
    }

    public GroundItem groundItemAt(Position position) {
        for (int i = groundItems.size() - 1; i >= 0; i--) {
            if (groundItems.get(i).position().equals(position)) {
                return groundItems.get(i);
            }
        }
        return null;
    }

    public void removeGroundItem(GroundItem item) {
        groundItems.remove(item);
    }

    /** Advances respawn timers, burns down fires and despawns dropped items. */
    public void tick() {
        Iterator<Map.Entry<Position, WorldObject>> objectIterator = objects.entrySet().iterator();
        while (objectIterator.hasNext()) {
            if (!objectIterator.next().getValue().tick()) {
                objectIterator.remove();
            }
        }
        groundItems.removeIf(item -> !item.tick());
    }

    /** The nearest walkable tile to {@code target}, used when the destination itself is blocked. */
    public Position nearestWalkable(Position target, int maxRadius) {
        if (isWalkable(target)) {
            return target;
        }
        for (int radius = 1; radius <= maxRadius; radius++) {
            for (int dy = -radius; dy <= radius; dy++) {
                for (int dx = -radius; dx <= radius; dx++) {
                    if (Math.max(Math.abs(dx), Math.abs(dy)) != radius) {
                        continue;
                    }
                    Position candidate = target.translate(dx, dy);
                    if (isWalkable(candidate)) {
                        return candidate;
                    }
                }
            }
        }
        return null;
    }
}

File: src/game/world/world.ts
import { Npc } from "../npc/npc";
import { GroundItem, WorldObject } from "./objects";
import { Position, posKey, samePosition, translate } from "./position";
import { TILE_INFO, TileType } from "./tiles";

/** The tile grid plus everything standing on it. */
export class World {
  readonly objects = new Map<string, WorldObject>();
  readonly npcs: Npc[] = [];
  readonly groundItems: GroundItem[] = [];

  constructor(
    readonly name: string,
    readonly width: number,
    readonly height: number,
    private readonly tiles: TileType[][],
    readonly spawn: Position,
  ) {}

  inBounds(x: number, y: number): boolean {
    return x >= 0 && y >= 0 && x < this.width && y < this.height;
  }

  tile(x: number, y: number): TileType {
    return this.inBounds(x, y) ? this.tiles[y][x] : "wall";
  }

  tileAt(position: Position): TileType {
    return this.tile(position.x, position.y);
  }

  isWalkable(position: Position): boolean {
    const { x, y } = position;
    if (!this.inBounds(x, y) || !TILE_INFO[this.tiles[y][x]].walkable) return false;
    const object = this.objects.get(posKey(position));
    return !object || !object.blocksMovement();
  }

  addObject(object: WorldObject): void {
    this.objects.set(posKey(object.position), object);
  }

  objectAt(position: Position): WorldObject | null {
    return this.objects.get(posKey(position)) ?? null;
  }

  removeObject(position: Position): void {
    this.objects.delete(posKey(position));
  }

  npcAt(position: Position): Npc | null {
    return this.npcs.find((npc) => npc.isVisible() && samePosition(npc.position, position)) ?? null;
  }

  addGroundItem(item: GroundItem): void {
    this.groundItems.push(item);
  }

  groundItemAt(position: Position): GroundItem | null {
    for (let i = this.groundItems.length - 1; i >= 0; i--) {
      if (samePosition(this.groundItems[i].position, position)) return this.groundItems[i];
    }
    return null;
  }

  removeGroundItem(item: GroundItem): void {
    const index = this.groundItems.indexOf(item);
    if (index >= 0) this.groundItems.splice(index, 1);
  }

  /** Advances respawn timers, burns fires down and despawns dropped loot. */
  tick(): void {
    for (const [key, object] of this.objects) {
      if (!object.tick()) this.objects.delete(key);
    }
    for (let i = this.groundItems.length - 1; i >= 0; i--) {
      if (!this.groundItems[i].tick()) this.groundItems.splice(i, 1);
    }
  }

  /** The closest walkable tile to a blocked destination. */
  nearestWalkable(target: Position, maxRadius: number): Position | null {
    if (this.isWalkable(target)) return target;
    for (let radius = 1; radius <= maxRadius; radius++) {
      for (let dy = -radius; dy <= radius; dy++) {
        for (let dx = -radius; dx <= radius; dx++) {
          if (Math.max(Math.abs(dx), Math.abs(dy)) !== radius) continue;
          const candidate = translate(target, dx, dy);
          if (this.isWalkable(candidate)) return candidate;
        }
      }
    }
    return null;
  }
}
